### PR TITLE
Add unicode fast path to `is_printable`

### PR DESCRIPTION
**Pull request by Noratrieb**
Before, it would enter the full expensive check even for normal ascii characters. Now, it skips the check for the ascii characters in `32..127`. This range was checked manually from the current behavior.

I ran the `tracing` test suite in miri, and it was really slow. I looked at a profile, and miri spent most of the time in `core::char::methods::escape_debug_ext`, where half of that was dominated by `core::unicode::printable::is_printable`. So I optimized it here.

The tracing profile:
The tracing profile

### CODE REVIEW

**Comment by Nilstrieb** [comment]
I've added a new benchmark and ran and old one locally, the results are promising. (`write_str_macro_debug` writes a single unicode character, and `write_str_macro_debug_ascii` writes several ascii characters to a `String`.)

It appears to be slightly slower for non-ascii, and a lot faster for ascii.
```
-- unicode-is-printable-fastpath run 1
test fmt::write_str_macro_debug                                    ... bench:     472,941 ns/iter (+/- 10,715)
test fmt::write_str_macro_debug_ascii                              ... bench:     213,816 ns/iter (+/- 17,440)

-- master run 1
test fmt::write_str_macro_debug                                    ... bench:     470,989 ns/iter (+/- 30,310)
test fmt::write_str_macro_debug_ascii                              ... bench:     352,146 ns/iter (+/- 18,440)

-- unicode-is-printable-fastpath run 2
test fmt::write_str_macro_debug                                    ... bench:     473,699 ns/iter (+/- 20,190)
test fmt::write_str_macro_debug_ascii                              ... bench:     213,746 ns/iter (+/- 11,796)

-- master run 2
test fmt::write_str_macro_debug                                    ... bench:     470,464 ns/iter (+/- 16,985)
test fmt::write_str_macro_debug_ascii                              ... bench:     348,037 ns/iter (+/- 14,689)
```

File: library/core/src/unicode/printable.py [hunk 1]
```diff
@@ -211,7 +211,11 @@ def main():
 pub(crate) fn is_printable(x: char) -> bool {
     let x = x as u32;
     let lower = x as u16;
-    if x < 0x10000 {
+
+    if lower >= 32 && lower < 127 {
+        // ASCII fast path
+        true
+    } else if x < 0x10000 {
```
Comment (joshtriplett) [inline library/core/src/unicode/printable.py:218]: I don't think this is quite correct. `lower` is just the lower 16 bits, so this conditional would also match any character whose lower 16 bits is in 32..=126, even if the upper bits are non-zero. And as far as I can tell, I don't think that's actually true for all of Unicode; for instance, `U+E0020` isn't printable.

Comment (joshtriplett) [inline library/core/src/unicode/printable.py:218]:
```suggestion

    if x < 32 {
        // ASCII fast path
        false
    } else if x < 127 {
        // ASCII fast path
        true
    } else if x < 0x10000 {
```

Comment (Nilstrieb) [reply library/core/src/unicode/printable.py]: Good catch, that is definitely wrong

**Comment by joshtriplett** [comment]
Since the conditional is needed anyway, would it be worth returning false in the < 32 case, to include that in the fast path?

**Comment by Nilstrieb** [comment]
Huh, I changed this locally but forgot to commit it.

**Pull request merged**